ic-factory/src/state.rs

```diff
@@ -456,76 +461,82 @@ impl Default for FactoryConfiguration {
     }
 }
 
-// The canister creation fee is 10^11 cycles, so we require the provided amount to be a little larger.
-// According to IC docs, 10^12 cycles should always cost 1 SDR, with is ~$1.
-const MIN_CANISTER_CYCLES: u64 = 10u64.pow(12);
+// The canister creation fee is 10^12 cycles, so we require the provided amount to be a little larger for the operation for the cansiter.
+// According to IC docs, 10^12 cycles should always cost 1 XDR, with is ~$1.
+const MIN_CANISTER_CYCLES: u64 = 10u64.pow(12) * 5;
 
-async fn consume_provided_cycles_or_icp(
+async fn consume_provided_icp(
     caller: Principal,
     ledger: Principal,
     icp_to: Principal,
     icp_fee: u64,
     controller: Principal,
 ) -> Result<u64, FactoryError> {
-    if ic_kit::ic::msg_cycles_available() > 0 {
-        consume_message_cycles()
-    } else {
-        if caller != controller {
-            consume_provided_icp(caller, ledger, icp_to, icp_fee).await?;
-        }
-
-        Ok(MIN_CANISTER_CYCLES)
-    }
-}
-
-fn consume_message_cycles() -> Result<u64, FactoryError> {
-    let amount = ic_kit::ic::msg_cycles_available();
-    if amount < MIN_CANISTER_CYCLES {
-        return Err(FactoryError::NotEnoughCycles(amount, MIN_CANISTER_CYCLES));
+    if caller != controller {
+        // If the caller is not the controller, we require the caller to provide cycles.
+        return transfer_and_top_up(icp_fee, ledger, caller, icp_to).await;
     }
 
-    Ok(ic_kit::ic::msg_cycles_accept(amount))
+    Ok(MIN_CANISTER_CYCLES)
 }
 
-async fn consume_provided_icp(
-    caller: Principal,
+/// Converts the `MIN_CANISTER_CYCLES` to ICP tokens, and the caller send
+/// the tokens to the cycles-minting-canister, the factory canister
+/// is topped up with cycles and the the icp_fee is sent to the
+/// `icp_to` principal.
+async fn transfer_and_top_up(
+    icp_fee: u64,
     ledger: Principal,
+    caller: Principal,
     icp_to: Principal,
-    icp_fee: u64,
-) -> Result<(), FactoryError> {
+) -> Result<u64, FactoryError> {
     let id = ic_kit::ic::id();
     let balance = ledger
         .get_balance(id, Some((&PrincipalId(caller)).into()))
         .await
         .map_err(FactoryError::LedgerError)?;
 
-    if balance < icp_fee + DEFAULT_TRANSFER_FEE.get_e8s() {
-        return Err(FactoryError::NotEnoughIcp(
-            balance,
-            icp_fee + DEFAULT_TRANSFER_FEE.get_e8s(),
-        ));
+    // defensive programming, maximum of twice the icp_fee
+    let top_up_fee = top_up::cycles_to_icp(MIN_CANISTER_CYCLES)
+        .await?
+        .min(icp_fee);
+
+    if balance < top_up_fee + icp_fee {
+        Err(FactoryError::NotEnoughIcp(balance, top_up_fee + icp_fee))?;
     }
 
-    consume_icp(caller, icp_fee, icp_to, ledger).await?;
+    let block_height =
+        top_up::transfer_icp_to_cmc(top_up_fee, ledger, Subaccount::from(&PrincipalId(caller)))
+            .await?;
```

Comment:
I have a feeling that some kind of error sneaked through here:
* `icp_fee` is the amount of ICP to be sent to the `icp_to` principal just as a commission for the operation
* `top_up_fee` is the amount of ICP to be converted to cycles for the factory
If we don't want to have any extra commission for creating canisters, `icp_fee` can be set to 0. But in that case `top_up_fee` will also be equal to 0, so no cycles will be added to the factory. This will also mean that if `icp_fee` is set below amount corresponding to `MIN_CANISTER_CYCLES`, then the canister will be created with less than `MIN_CANISTER_CYCLES` which sounds strange. It's even possible for the canister creation to fail if the cycles amount is to little in the end.

Please, check that this logic works as intended.